### PR TITLE
Improve performance during CheckAttesterDoubleVotes

### LINKED ISSUE #8913: Optimize `CheckAttesterDoubleVotes` Slasher DB Method
# 💎 Issue

### Background

Our optimized slasher implementation is a total rewrite of how slasher is meant to function. Although it is meant to be more efficient from first principles, there are a few low-hanging fruit we want to improve to ensure the implementation is up to high standards.

One of the slow functions is `CheckAttesterDoubleVotes` found https://github.com/prysmaticlabs/prysm/blob/401f73d341a119feccacb1f9cf56d11d2b9fdea6/beacon-chain/db/slasherkv/slasher.go#L89. The goal of the function is to detect if a validator has created a double vote attestation, that is, the validator has cast two attestations at the same target epoch with different attestation data roots.

The current implementation is fairly naive, doing the following:
```
Take in a list of attestations
Create an empty list of double votes found
for each attestation:
  for each validator in its attesting indices:
    database_key = attestation.target_epoch ++ validator
    if exists_in_db(database_key): append to list of double votes
```

We could have hundreds of attesting indices and hundreds of attestations passed into this function. The impact on CPU is quite clear here: 

Screen_Shot_2021-05-20_at_9 44 50_AM

### Description

We want to identify a solution that improves this performance bottleneck and can be benchmarked.

### PR DESCRIPTION
**Pull request by jmozah**
**What type of PR is this?**
Performance Improvement

**What does this PR do? Why is it needed?**
This PR tries to improves the performance when checking for double votes during attestation.
The code was doing multiple in-directions inside a double loop.
The fix was to store the attestations in multiple places (de-normalize) and process
every attestation in a seperate go routine.

**Which issues(s) does this PR fix?**
Fixes #8913 

**Other notes for review**
- Bucket removed (attestationRecordsBucket)
- Bucket 'attestationDataRootsBucket' value changed from 
  'att.SigningRoot + FastSum64([]ValidatorIndex)' to 'att.SigningRoot + AttestationRecord'
- DB migration after merging 'feature/slasher' ??

### CODE REVIEW

**Comment by jmozah** [comment]
### Benchmarks: Old schema vs New shema
BenchmarkStore_CheckDoubleBlockProposals-16            2         504269674 ns/op        208081248 B/op   5010016 allocs/op
BenchmarkStore_CheckDoubleBlockProposals-16            7         157626795 ns/op        189470513 B/op   4086973 allocs/op

File: beacon-chain/db/slasherkv/schema.go [hunk 1]
```diff
@@ -9,7 +9,6 @@ package slasherkv
 var (
 	// Slasher buckets.
 	attestedEpochsByValidator  = []byte("attested-epochs-by-validator")
-	attestationRecordsBucket   = []byte("attestation-records")
```
Comment (prestonvanloon) [inline beacon-chain/db/slasherkv/schema.go:12]: We should add a database "migration" to delete this bucket to reclaim the space, if it is no longer used. 

@rauljordan, what do you think?

File: beacon-chain/db/slasherkv/slasher.go [hunk 2]
```diff
@@ -92,43 +93,55 @@ func (s *Store) CheckAttesterDoubleVotes(
...
-				attRecordsKey := signingRootsBkt.Get(validatorEpochKey)
-
-				// An attestation record key is comprised of a signing root (32 bytes)
-				// and a fast sum hash of the attesting indices (8 bytes).
-				if len(attRecordsKey) < attestationRecordKeySize {
-					continue
+	doubleVotesMu := sync.Mutex{}
+	eg, egctx := errgroup.WithContext(ctx)
+	for _, att := range attestations {
+		attToProcess := att  // https://golang.org/doc/faq#closures_and_goroutines
```
Comment (prestonvanloon) [inline beacon-chain/db/slasherkv/slasher.go:99]: It is not clear how this comment adds additional context to the reader. Could you please use a full sentence followed by "See: $URL".

**Review by rauljordan — DISMISSED** [review]
Wow! Really great solution and thank you for the benchmark and detailed comments. I think adding errgroup is also a great way to speed up given just how many attestations this method is called with. Great work

File: beacon-chain/db/slasherkv/slasher.go [hunk 3]
```diff
@@ -92,43 +93,55 @@ func (s *Store) CheckAttesterDoubleVotes(
...
+		eg.Go(func() error { // process every attestation parallely
```
Comment (prestonvanloon) [inline beacon-chain/db/slasherkv/slasher.go:100]:
```suggestion
		eg.Go(func() error { // process every attestation parallelly
```

File: beacon-chain/db/slasherkv/slasher.go [hunk 4]
```diff
@@ -361,18 +361,17 @@ func (s *Store) HighestAttestations(
...
+					encodedAttRecord := v[signingRootSize:]
 					if encodedAttRecord == nil {
 						continue
 					}
-					attWrapper, err := decodeAttestationRecord(encodedAttRecord)
-					if err != nil {
-						return err
+					attWrapper, err1 := decodeAttestationRecord(encodedAttRecord)
+					if err1 != nil {
+						return err1
```
Comment (prestonvanloon) [inline beacon-chain/db/slasherkv/slasher.go:374]: I'm curious about this change. Why rename err to err1? Was it to satisfy some err / shadowing linter, personal preference, or some other standard we should adopt?

**Review by prestonvanloon — COMMENTED** [review]
Very nice PR! I had just a few comments or questions.

File: beacon-chain/db/slasherkv/schema.go (hunk 1 above)
Comment (rauljordan) [reply beacon-chain/db/slasherkv/schema.go:12]: This db is unused at the moment, so we are safe, but we should add a migration in the future for slasher changes

Comment (jmozah) [reply beacon-chain/db/slasherkv/schema.go:12]: I had the same question. But @rauljordan clarified that we do migration in another PR as not many people use it

File: beacon-chain/db/slasherkv/slasher.go (hunk 4 above)
Comment (jmozah) [reply beacon-chain/db/slasherkv/slasher.go:374]: Yes. err is returned by View. By naming like this we can avoid shadowing. We should actually use more legible errName. I have changed this to 'decodeErr'

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**